napari/utils/_logging.py

```diff
@@ -0,0 +1,49 @@
+import logging
+
+# keep track of all the logging info in a stream
+_LOG_SEPARATOR = '<NAPARI_LOG_SEPARATOR>'
+
+
+class _LogHandler:
+    def __init__(self):
+        self.logs = []
+
+    def write(self, log_msg):
+        level_name, time, thread, msg = log_msg.split(_LOG_SEPARATOR)
+        levels = logging.getLevelNamesMapping()
+        level_value = levels[level_name]
+        self.logs.append((level_value, level_name, time, thread, msg))
+
+    def flush(self):
+        pass
+
+    def __str__(self):
+        return ''.join(
+            [
+                f'[{time}] ({thread}) {level_name}: {msg}'
+                for _, level_name, time, thread, msg in self.logs
+            ]
+        )
+
+    def logs_at_level(self, level=logging.DEBUG):
+        if isinstance(level, str):
+            level = logging.getLevelName(level)
+        return [
+            (level_value, *others)
+            for level_value, *others in self.logs
+            if level_value >= level
+        ]
+
+
+def _get_custom_log_stream():
+    log_stream = _LogHandler()
+    handler = logging.StreamHandler(log_stream)
+    handler.setFormatter(
+        logging.Formatter(
+            f'%(levelname)s{_LOG_SEPARATOR}%(asctime)s{_LOG_SEPARATOR}%(threadName)s{_LOG_SEPARATOR}%(message)s'
+        )
+    )
+    logger = logging.getLogger('napari')
+    logger.setLevel(logging.DEBUG)
+    logger.addHandler(handler)
```

Comment:
It will be nice to connect to all loggers and be able to filter messages based on logger (so someone could filter logs only from a given plugin logger.

Reply:
True... I'm not sure how to do that though; I guess we somehow get the root logger? And btw, is there a way to get these logging parameters (level, thread, etc) "raw" instead of formatting to string and then de-formatting?

Reply:
EDIT: I changed to use the root logger and filter for `napari` by default. Probably less finnicky than registering specific plugins.